Remove FindLocalVal

hamishknight opened the pull request:
Replace with an ASTScope lookup. This also lets us simplify UsableFilteringDeclConsumer, and fixes a couple of completion bugs.

Resolves #66925
Resolves #71384
rdar://92479209
rdar://108164277
rdar://122302481

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

slavapestov commented:
Yes!

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -1171,167 +1171,131 @@ static void lookupVisibleMemberDecls(
+void swift::lookupVisibleDecls(VisibleDeclConsumer &ParentConsumer,
+                               SourceLoc Loc, const DeclContext *DC,
+                               bool IncludeTopLevel) {
+  auto *SF = DC->getParentSourceFile();
+  assert(SF);
+  assert(Loc.isValid());
```
Any objections to requiring a valid SourceLoc and SourceFile? We could potentially fall back to a module-scope lookup, but it's not clear to me how useful such a behavior would be (seems like callers ought to use `lookupVisibleDeclsInModule` if that's what they really want), and it seems like all the current clients meet this requirement.

rintaro replied:
No objection. Maybe we should even add
```
assert(DC->isModuleScopeContext());
```
because without `IncludeTopLevel: true`, that lookup doesn't make any sense. If the caller really wants _only_ top-level decls, they usually should use `ModuleDecl::lookupVisibleDecls()`

hamishknight replied:
Presumably you mean `assert(!DC->isModuleScopeContext())`? Seems reasonable enough to me

hamishknight replied:
Presumably you mean `assert(!DC->isModuleScopeContext());`? Looks like that can be hit for e.g decl attribute completion, and I think seems reasonable to have a module scope context there

rintaro replied:
OK. But just out of curiosity, does `swift::lookupVisibleDecls()` at `SourceFile` context without `IncludeTopLevel` emit anything?

hamishknight replied:
No it doesn't include anything

hamishknight commented:
Same stress tester failures as main, except for `TodayTurnipSection.swift at offset 1405`, which is a soft timeout on main (although it's at 293 seconds, so it's pretty close to exceeding the 300 second timeout as-is, and testing locally I get a ~1s speedup with this patch from ~55s to ~54s).

hamishknight commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -1171,167 +1171,131 @@ static void lookupVisibleMemberDecls(
-      ExtendedType = ED->getSelfTypeInContext();
-    } else if (auto ND = dyn_cast<NominalTypeDecl>(DC)) {
-      ExtendedType = ND->getSelfTypeInContext();
+      VD->visitAuxiliaryDecls([&](Decl *auxDecl) {
+        if (auto *auxValue = dyn_cast<ValueDecl>(auxDecl))
+          foundDecl(auxValue);
+      });
```
Seems like ASTScopeDeclConsumerForLocalLookup currently just doesn't handle peer macros 🤔

slavapestov replied:
The goal is to eventually rewrite ASTScope to look at swift-syntax only, so semantic members should be introduced by clients (but we can factor out a common client for everyone to use that also visits semantic local members).

rintaro replied:
Okay,

Previous `FindLocalVal` implementation handling aux decls was recursive.
```
  std::function<void(Decl *)> visitDecl;
  visitDecl = [&](Decl *D) {
    if (auto *VD = dyn_cast<ValueDecl>(D))
      checkValueDecl(VD, DeclVisibilityKind::LocalVariable);
    D->visitAuxiliaryDecls(visitDecl);
  };
  ...
  visitDecl(D);
```
This was to handle multi level macro expansion (e.g. `peer` macro expand to a decl with another `peer` macro) I _think_ we need similar treatment.

hamishknight replied:
After speaking offline, updated to match ASTScopeDeclConsumerForLocalLookup, since peer decls for local decls should not be included in the lookup result.

hamishknight commented:
@swift-ci please test

hamishknight committed:
NFC: Rename `DeclVisibilityKind::LocalVariable` -> `LocalDecl`

This includes local types, so make the name a bit
more generic.

hamishknight committed:
Remove FindLocalVal

Replace with an ASTScope lookup. This also lets
us simplify UsableFilteringDeclConsumer, and fixes
a couple of completion bugs.

hamishknight commented:
@swift-ci please test

rintaro reviewed (approved):
lgtm!
